[quant][graphmode] Support quantization for `aten::apend`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40889 [quant][graphmode][fix] Print the node in error message
* #40865 [quant][graphmode][fix] Fold conv bn
* **#40743 [quant][graphmode] Support quantization for `aten::apend`**

Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22302151](https://our.internmc.facebook.com/intern/diff/D22302151)

jerryzh168 committed:
[quant][graphmode] Support quantization for `aten::apend`

Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented:
This changes the underlying behavior: We are consuming more memory than before, is there a way to support append directly? How is the behavior of append different from that of cat?

jerryzh168 commented:
> This changes the underlying behavior: We are consuming more memory than before, is there a way to support append directly? How is the behavior of append different from that of cat?

`aten::append` is inplace, and the output is ignored. If script/trace can generate a non inplace `append` then we'll be able to quantize `append` directly, similar to concat

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -1100,6 +1100,49 @@ void InsertObserversHelper::fillBoundaryValueMap(
 }
 
+void makeAppendNonInplace(std::shared_ptr<Graph>& graph) {
+  std::string append_pattern = R"IR(
```
Can you elaborate more on why directly supporting append is not possible?

jerryzh168 replied:
All current ops including inplace ops assumes that the output will be consumed by the following ops. To break this assumption, we'll need to introduce substantial changes/hacks, I don't think it's worth the effort

jerryzh168 replied:
btw, if there are perf problems we can easily add a pass to change the add to append in the end

raghuramank100 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -309,10 +309,16 @@ std::vector<Value*> getPassThroughInputs(Value* v) {
-  } else if (n->kind() == Symbol::aten("append")) {
-    // notice that append is an op that changes input inplace
-    return {n->input(0), n->input(1)};
+  } else if (isListAdd(n)) {
```
What is the rule for list add? Do we assume that all the tensors in the list have a dequantize op prior to the add?

jerryzh168 replied:
this is the transformed "append", as shown in the description. we will check if the inputs are produced with dequantize to make sure the inputs are quantized

jerryzh168 replied:
the rule of list add:
```
%y = aten::add(%list, %x)
```
we'll check if %list is empty list, if it is, then the pass through list for %y is {%x}, otherwise, it is {%list, %x}

raghuramank100 replied:
How do you check if the existing list has only dequantized tensors?. I see how you can do it for the input %x, dont follow how the check is done for %list

jerryzh168 replied:
There are two cases for %list:
1. 
one case is same as %x
```
%list = aten::dequantize(%list_quant)
%result = aten::add(%list, %x_list)
```
we check if the list is produced by dequantize or not
2.
another case is when list is empty, it can be considered as containing quantized tensors
```
%list : Tensor[] = prim::ListConstruct()
%result = aten::add(%list, %x_list)
```
in this case we only need to check if %x_list is produced by dequantize.

raghuramank100 replied:
My question is on how we check if %list is produced by dequantize or not. Do we iterate over all elements in the list?

jerryzh168 replied:
no we don't, we just check if %list is produced by dequantize or not

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@raghuramank100 I added a fix to change add back to append in finalize, please take a look again.

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22302151](https://our.internmc.facebook.com/intern/diff/D22302151)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22302151](https://our.internmc.facebook.com/intern/diff/D22302151)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22302151](https://our.internmc.facebook.com/intern/diff/D22302151)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22302151](https://our.internmc.facebook.com/intern/diff/D22302151)

[ghstack-poisoned]

vkuzo commented:
btw, where is the source code for aten::append?  didn't find it in native_functions.yaml nor in the pytorch c++ docs (might have been looking in the wrong place)

vkuzo reviewed (approved):
accepting to unblock

jerryzh168 commented:
> btw, where is the source code for aten::append? didn't find it in native_functions.yaml nor in the pytorch c++ docs (might have been looking in the wrong place)

not sure it's probably something just in JIT

jerryzh168 committed:
Update on "[quant][graphmode] Support quantization for `aten::apend`"


Summary:
`aten::append` modifies input inplace and the output is ignored, these ops are not
supported right now, so we'll need to first make `aten::append` non-inplace
by change
```
ignored = aten::append(list, x)
```
to
```
x_list = aten::ListConstruct(x)
result = aten::add(list, x_list)
```
and then quantize the aten::add instead.

Test Plan:

TestQuantizeJitOps.test_general_shape_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22302151](https://our.internmc.facebook.com/intern/diff/D22302151)

[ghstack-poisoned]

jerryzh168 commented:
> btw, where is the source code for aten::append? didn't find it in native_functions.yaml nor in the pytorch c++ docs (might have been looking in the wrong place)

found it: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/runtime/register_prim_ops.cpp.html#362
